pkg/util/validation/validate.go

```diff
@@ -0,0 +1,117 @@
+package validation
+
+import (
+	"net/http"
+	"time"
+
+	"flag"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/cortex/pkg/ingester/client"
+	"github.com/weaveworks/cortex/pkg/util/extract"
+)
+
+const (
+	discardReasonLabel = "reason"
+
+	errMissingMetricName = "sample missing metric name"
+	errInvalidMetricName = "sample invalid metric name: %.200q"
+	errInvalidLabel      = "sample invalid label: %.200q metric %.200q"
+	errLabelNameTooLong  = "label name too long: %.200q metric %.200q"
+	errLabelValueTooLong = "label value too long: %.200q metric %.200q"
+	errTooManyLabels     = "sample for '%s' has %d label names; limit %d"
+	errTooOld            = "sample for '%s' has timestamp too old: %d"
+	errTooNew            = "sample for '%s' has timestamp too new: %d"
+
+	greaterThanMaxSampleAge = "greater_than_max_sample_age"
+	maxLabelNamesPerSeries  = "max_label_names_per_series"
+	tooFarInFuture          = "too_far_in_future"
+)
+
+// DiscardedSamples is a metric of the number of discarded samples, by reason.
+var DiscardedSamples = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "cortex_discarded_samples_total",
+		Help: "The total number of samples that were discarded.",
+	},
+	[]string{discardReasonLabel},
+)
+
+func init() {
+	prometheus.MustRegister(DiscardedSamples)
+}
+
+// Config for validation settings and options.
+type Config struct {
+	// maximum length a label name can be
+	MaxLabelNameLength int
+
+	// maximum length a label value can be. This also is the maximum length of a metric name.
+	MaxLabelValueLength int
+
+	// maximum number of label/value pairs timeseries.
+	MaxLabelNamesPerSeries int
+
+	// Config for rejecting old samples.
+	RejectOldSamples       bool
+	RejectOldSamplesMaxAge time.Duration
+	CreationGracePeriod    time.Duration
+}
+
+// RegisterFlags registers a set of command line flags for setting options regarding sample validation at ingestion time.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.IntVar(&cfg.MaxLabelNameLength, "validation.max-length-label-name", 1024, "Maximum length accepted for label names")
+	f.IntVar(&cfg.MaxLabelValueLength, "validation.max-length-label-value", 2048, "Maximum length accepted for label value. This setting also applies to the metric name")
+	f.IntVar(&cfg.MaxLabelNamesPerSeries, "validation.max-label-names-per-series", 20, "Maximum number of label names per series.")
```

Comment:
btw I now think this default is too low - we've got it set to 30

Reply:
Might as well bump it in this PR then, give us a sec.